src/backend/distributed/planner/multi_router_planner.c

```diff
@@ -687,23 +689,56 @@ ModifyQuerySupported(Query *queryTree, Query *originalQuery, bool multiShardQuer
 	}
 
 	/*
-	 * Reject queries which involve joins. Note that UPSERTs are exceptional for this case.
-	 * Queries like "INSERT INTO table_name ON CONFLICT DO UPDATE (col) SET other_col = ''"
-	 * contains two range table entries, and we have to allow them.
+	 * We have to allow modify queries with two range table entries, if it is pushdownable.
 	 */
 	if (commandType != CMD_INSERT && queryTableCount != 1)
 	{
-		/*
-		 * We support UPDATE and DELETE with joins unless they are multi shard
-		 * queries.
-		 */
-		if (!UpdateOrDeleteQuery(queryTree) || multiShardQuery)
+		/* We can not get restriction context via master_modify_multiple_shards path */
+		if (plannerRestrictionContext == NULL)
 		{
 			return DeferredError(ERRCODE_FEATURE_NOT_SUPPORTED,
-								 "cannot perform distributed planning for the given "
-								 "modification",
-								 "Joins are not supported in distributed "
-								 "modifications.", NULL);
+								 "cannot run multi shard modify query with master_modify_multiple_shard",
```

Comment:
Maybe `cannot run multi shard modify query with master_modify_multiple_shard when the query involves subquery or join` ? We generally try to give more useful information in the error messages.

`Send your query directly.` -> `Execute the query without using master_modify_multiple_shard()` ? (I'm not sure about this, maybe let's wait @marcocitus )